### PR TITLE
feat: [PIE-2913]: Creation of environment groups api 

### PR DESCRIPTION
**Pull request by prashantsharma04**
You can use the following comments to re-trigger PR Checks

- Compile: `trigger compile`
- runAeriformCheck: `trigger AeriformCheck`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/31438)

### CODE REVIEW

File: 125-cd-nextgen/src/main/java/io/harness/cdng/envGroup/beans/EnvironmentGroupConfig.java [hunk 1]
```diff
@@ -28,12 +30,12 @@
...
 
-  String orgIdentifier;
-  String projectIdentifier;
+  @NotNull @Trimmed String orgIdentifier;
+  @NotNull @Trimmed String projectIdentifier;
 
   String description;
   String color;
-  List<NGTag> tags;
+  Map<String, String> tags;
```
Comment (naidu-harness) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/envGroup/beans/EnvironmentGroupConfig.java:38]: why cant we use <NGTag>

File: 125-cd-nextgen/src/main/java/io/harness/repositories/envGroup/EnvironmentGroupRepositoryCustomImpl.java [hunk 2]
```diff
@@ -37,4 +50,31 @@
...
+        -> outboxService.save(EnvironmentGroupCreateEvent.builder()
+                                  .accountIdentifier(environmentGroupEntity.getAccountIdentifier())
+                                  .orgIdentifier(environmentGroupEntity.getOrgIdentifier())
+                                  .projectIdentifier(environmentGroupEntity.getProjectIdentifier())
+                                  .environmentGroupEntity(environmentGroupEntity)
+                                  .build());
+    try {
+      savedEntity = gitAwarePersistence.save(environmentGroupEntity, environmentGroupEntity.getYaml(), ChangeType.ADD,
+          EnvironmentGroupEntity.class, functor);
+    } catch (DuplicateKeyException ex) {
```
Comment (namanvermah) [inline 125-cd-nextgen/src/main/java/io/harness/repositories/envGroup/EnvironmentGroupRepositoryCustomImpl.java:67]: No need to catch duplicate key here if service layer is catching this

**Review by archit-harness — CHANGES_REQUESTED** [review]
> Reviewed 12 of 16 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/31438)* status: 16 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[125-cd-nextgen/src/main/java/io/harness/cdng/envGroup/beans/EnvironmentGroupConfig.java, line 38 at r1](https://reviewable.io/reviews/harness/harness-core/31438#-My7vWMmFNQdeJKB1dlS-r1-38:-MyGyQjV39Xzr3bG4sjf:b-qkepw9) ([raw file](https://github.com/harness/harness-core/blob/00ef70405915595828f8781303ce2a76d4217469/125-cd-nextgen/src/main/java/io/harness/cdng/envGroup/beans/EnvironmentGroupConfig.java#L38)):*
<details><summary><i>Previously, naidu-harness (Sanyasi Naidu Annepu) wrote…</i></summary><blockquote>

why cant we use `NGTag`
</blockquote></details>

List<NGTag> tags should use this
___
*[125-cd-nextgen/src/main/java/io/harness/cdng/envGroup/services/EnvironmentGroupServiceImpl.java, line 57 at r1](https://reviewable.io/reviews/harness/harness-core/31438#-MyGyUtN8k-MSqBIpWbm:-MyGyUtN8k-MSqBIpWbn:bfezdms) ([raw file](https://github.com/harness/harness-core/blob/00ef70405915595828f8781303ce2a76d4217469/125-cd-nextgen/src/main/java/io/harness/cdng/envGroup/services/EnvironmentGroupServiceImpl.java#L57)):*
> ```Java
>       throw new InvalidRequestException(
>           String.format("Error occurred while saving environment group %s", e.getMessage()));
>     }
> ```

write todo, and create ticket for enforcement if its valid for this



_Code quote:_
```Java
    try {
      return environmentRepository.create(entity);
    } catch (DuplicateKeyException ex) {
      throw new DuplicateFieldException(
          format("Identifier %s already exists in organization id %s and project id %s", entity.getIdentifier(),
              entity.getProjectIdentifier(), entity.getOrgIdentifier()),
          USER_SRE, ex);
    } catch (Exception e) {
      throw new InvalidRequestException(
          String.format("Error occurred while saving environment group %s", e.getMessage()));
    }
```
___
*[125-cd-nextgen/src/main/java/io/harness/repositories/envGroup/EnvironmentGroupRepositoryCustomImpl.java, line 58 at r1](https://reviewable.io/reviews/harness/harness-core/31438#-MyGyoAdCGoWgupoIloK:-MyGyoAdCGoWgupoIloL:b1wc8q4) ([raw file](https://github.com/harness/harness-core/blob/00ef70405915595828f8781303ce2a76d4217469/125-cd-nextgen/src/main/java/io/harness/repositories/envGroup/EnvironmentGroupRepositoryCustomImpl.java#L58)):*
> ```Java
>     EnvironmentGroupEntity savedEntity;
>     Supplier<OutboxEvent> functor = ()
>         -> outboxService.save(EnvironmentGroupCreateEvent.builder()
> ```

should log audits only if if git sync is disabled or if git sync is enabled (only for default branch)



_Code quote:_
```Java
outboxService.save
```

File: 125-cd-nextgen/src/main/java/io/harness/cdng/envGroup/beans/EnvironmentGroupConfig.java (hunk 1 above)
Comment (prashantsharma04) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/envGroup/beans/EnvironmentGroupConfig.java:38]: In Entity, we have defined them as NGTag. This is the config file which will read from yaml. While converting to config to entities we are making it as NGTag.  (EnvironmentGroupMapper::toEnvironmentEntity)

I have handled the same way we are handling for environments

**Comment by prashantsharma04** [comment]
trigger messagecheck

**Review by prashantsharma04 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/31438)* status: 17 file(s) to review. 1 change requests. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[125-cd-nextgen/src/main/java/io/harness/cdng/envGroup/beans/EnvironmentGroupConfig.java, line 38 at r1](https://reviewable.io/reviews/harness/harness-core/31438#-My7vWMmFNQdeJKB1dlS-r1-38:-MyHfj6DDAfZ18AkhgNE:bim2das) ([raw file](https://github.com/harness/harness-core/blob/00ef70405915595828f8781303ce2a76d4217469/125-cd-nextgen/src/main/java/io/harness/cdng/envGroup/beans/EnvironmentGroupConfig.java#L38)):*
<details><summary><i>Previously, prashantsharma04 (Prashant Sharma) wrote…</i></summary><blockquote>

In Entity, we have defined them as NGTag. This is the config file which will read from yaml. While converting
 config to entities we are making it as NGTag.  (EnvironmentGroupMapper::toEnvironmentEntity)

I have handled the same way we are handling for environments
</blockquote></details>

@archit-harness As discussed - we do have list of NGTag in EnvGroupEntity
___
*[125-cd-nextgen/src/main/java/io/harness/repositories/envGroup/EnvironmentGroupRepositoryCustomImpl.java, line 58 at r1](https://reviewable.io/reviews/harness/harness-core/31438#-MyGyoAdCGoWgupoIloK:-MyHfts2Ec4_BdqpO8R-:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/00ef70405915595828f8781303ce2a76d4217469/125-cd-nextgen/src/main/java/io/harness/repositories/envGroup/EnvironmentGroupRepositoryCustomImpl.java#L58)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

should log audits only if if git sync is disabled or if git sync is enabled (only for default branch)
</blockquote></details>

Done.
___
*[125-cd-nextgen/src/main/java/io/harness/repositories/envGroup/EnvironmentGroupRepositoryCustomImpl.java, line 67 at r1](https://reviewable.io/reviews/harness/harness-core/31438#-My7vWMmFNQdeJKB1dlk-r1-67:-MyHfuc536MjBTnsDqHZ:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/00ef70405915595828f8781303ce2a76d4217469/125-cd-nextgen/src/main/java/io/harness/repositories/envGroup/EnvironmentGroupRepositoryCustomImpl.java#L67)):*
<details><summary><i>Previously, namanvermah (Naman Verma) wrote…</i></summary><blockquote>

No need to catch duplicate key here if service layer is catching this
</blockquote></details>

Done.

**Review by naidu-harness — APPROVED** [review]
> Reviewed 13 of 16 files at r1, 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/31438)* status: 6 file(s) to review. 1 change requests. 2. unresolved discussions. 1 of 2 approvals obtained.

**Review by archit-harness — CHANGES_REQUESTED** [review]
> Reviewed 2 of 16 files at r1, 3 of 4 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/31438)* status: 1 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[120-ng-manager/config.yml, line 241 at r3](https://reviewable.io/reviews/harness/harness-core/31438#-MyLFp0_Bq7je_C9Z9oY:-MyLFp0_Bq7je_C9Z9oZ:b4wldth) ([raw file](https://github.com/harness/harness-core/blob/8964311f616ac219f62b3fc2b7f87a9b3054c4e8/120-ng-manager/config.yml#L241)):*
> ```YAML
>   authority: localhost
> 
> shouldConfigureWithPMS: true
> ```

revert this



_Code quote:_
```YAML
true
```

**Review by prashantsharma04 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/31438)* status: 2 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[120-ng-manager/config.yml, line 241 at r3](https://reviewable.io/reviews/harness/harness-core/31438#-MyLFp0_Bq7je_C9Z9oY:-MyLKZ5iA6NeEnGiVCEL:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/8964311f616ac219f62b3fc2b7f87a9b3054c4e8/120-ng-manager/config.yml#L241)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

revert this
</blockquote></details>

Done.

**Review by archit-harness — APPROVED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/31438)* status: 2 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by sahilhindwani-harness — APPROVED** [review]
> Reviewed 13 of 16 files at r1, 2 of 4 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/31438)* status: 1 file(s) to review. 1. unresolved discussions.

**Review by puthrayaharness — APPROVED** [review]
> Reviewed 13 of 16 files at r1, 2 of 4 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/31438)* status: 1 file(s) to review. 1. unresolved discussions.

**Review by namanvermah — APPROVED** [review]
:lgtm:

> Reviewed 13 of 16 files at r1, 2 of 4 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/31438#-:-Myept4I7xQJw1BkDwlP:bnfp4nl)* status: :shipit: complete! ✓

**Pull request merged**